### DISCUSSION TITLE
How to build different capacity of SPI Flash for RP2 PICO?

### DISCUSSION THREAD

**Discussion by water5** [micropython/micropython#11739] [RP2040 / Pico] [answered]
I try to replace RP2 PICO SPI Flash to 16MB (W25Q128JV_SQ, this model already in `micropython/drivers/memory/external_flash_device.h`), and modify `micropython/ports/rp2/boards/PICO/mpconfigboard.h`, 

`#define MICROPY_HW_FLASH_STORAGE_BYTES          (1408 * 1024)`
to
`#define MICROPY_HW_FLASH_STORAGE_BYTES          (15 * 1024 * 1024)`

When build it, appear error:

> static_assert(MICROPY_HW_FLASH_STORAGE_BYTES <= PICO_FLASH_SIZE_BYTES, "MICROPY_HW_FLASH_STORAGE_BYTES too big");
>       | ^~~~~~~~~~~~~
> make[3]: *** [CMakeFiles/firmware.dir/build.make:3265: CMakeFiles/firmware.dir/rp2_flash.c.obj] Error 1
> make[2]: *** [CMakeFiles/Makefile2:1575: CMakeFiles/firmware.dir/all] Error 2
> make[1]: *** [Makefile:91: all] Error 2
> See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
> make: *** [Makefile:46: all] Error 1

it seems not work(MicroPython runs fine and prompt appear correct, but the MicroPython SPI Flash user drive not appear.) 

Which files should I modify?

**Comment by robert-hh** [comment]
You have to define PICO_FLASH_SIZE_BYTES in mpconfigboard.h as well as 16 * 1024 * 1024. It is by default defined as 2 MB.

**Reply by water5** [reply]
I append `#define PICO_FLASH_SIZE_BYTES (16 * 1024 * 1024)` to `mpconfigboard.h` and build it, appear error:

> micropython/ports/rp2/boards/PICO/mpconfigboard.h:4: error: "PICO_FLASH_SIZE_BYTES" redefined [-Werror]
    4 | #define PICO_FLASH_SIZE_BYTES     (16 * 1024 * 1024)
>
> ...
>
> micropython/lib/pico-sdk/src/boards/include/boards/pico.h:74: note: this is the location of the previous definition
   74 | #define PICO_FLASH_SIZE_BYTES (2 * 1024 * 1024)
>
> ...
>
> make[2]: *** [CMakeFiles/Makefile2:1575: CMakeFiles/firmware.dir/all] Error 2
> make[1]: *** [Makefile:91: all] Error 2
> See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
> make: *** [Makefile:46: all] Error 1

So it should be modify `micropython/lib/pico-sdk/src/boards/include/boards/pico.h` 
#define PICO_FLASH_SIZE_BYTES (16 * 1024 * 1024)

and build correct.

But what mean `micropython/lib/pico-sdk/src/boards/include/boards/pico.h` line 67:

`#define PICO_BOOT_STAGE2_CHOOSE_W25Q080 1`

although I not to modify this line, build pass and works fine, actually I use W25Q128.

**Reply by robert-hh** [reply, marked as answer]
You could indeed change the line in Pico-SDK, or `#undef PICO_FLASH_SIZE_BYTES` before re-defining it, or use a Board with 16 MB flash as template.

> But what mean micropython/lib/pico-sdk/src/boards/include/boards/pico.h line 67:

Flash chips may need slightly different command sets. So line defines, which flash chip (Windbond, 1 MB) is used. Which does not match the PICO anyhow. But most flash chips use the same command set for standard operations like read, program, erase.
